README.md
# line-profiler-extended

[![PyPI](https://img.shields.io/pypi/v/line-profiler-extended?style=flat-square)](https://pypi.python.org/pypi/line-profiler-extended/)
[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/line-profiler-extended?style=flat-square)](https://pypi.python.org/pypi/line-profiler-extended/)
[![PyPI - License](https://img.shields.io/pypi/l/line-profiler-extended?style=flat-square)](https://pypi.python.org/pypi/line-profiler-extended/)
[![Coookiecutter - Wolt](https://img.shields.io/badge/cookiecutter-Wolt-00c2e8?style=flat-square&logo=cookiecutter&logoColor=D4AA00&link=https://github.com/woltapp/wolt-python-package-cookiecutter)](https://github.com/woltapp/wolt-python-package-cookiecutter)


---

Inherits awesome rkern's line-profiler and adds some useful features.

## Installation

```sh
pip install line-profiler-extended
```

## Using the API

```python
from line_profiler_extended import LineProfilerExtended

def foo():
    pass

# profile the foo function
profiler = LineProfilerExtended(foo)

# profile all functions from some_module
import some_module
profiler = LineProfilerExtended(some_module)

# profile all functions from all modules found recursively
# starting from the grandparent directory of the current file
from pathlib import Path
profiler = LineProfilerExtended(Path(__file__).parent.parent)

# profile all functions from all modules found recursively in "path",
# reporting only functions that took at least 1 millisecond
profiler = LineProfilerExtended("path", eps=0.001)

# profile all functions from all modules found recursively in "path" with "m" in module name but without "mm"
profiler = LineProfilerExtended("path", include_regex="m", exclude_regex="mm")

# all types of locations can be combined
profiler = LineProfilerExtended(
    Path("/some/path"), "path", some_module, foo,
    eps=0.001, include_regex="m", exclude_regex="mm"
)

profiler.enable_by_count()
profiler.runcall(foo)
profiler.print_stats()
```

## Usage with IPython

```ipython
%load_ext line_profiler_extended

# profile the foo function
%lpext -p foo foo()

# profile all functions from some_module
%lpext -p some_module foo()

# profile all functions from all modules found recursively in some path
from pathlib import Path
%lpext -p Path("/some/path") foo()

# profile all functions from all modules found recursively in "path",
# reporting only functions that took at least 1 millisecond
%lpext -p "path" --eps 0.001 foo()

# profile all functions from all modules found recursively in "path" with "m" in module name but without "mm"
%lpext -p "path" --include "m" --exclude "mm" foo()

# all types of locations can be combined
%lpext -p Path(__file__).parent.parent -p "path" -p some_module -p foo --eps 0.001 --include "m" --exclude "mm" foo()
```

## Usage with pytest

```python
import pytest
from pathlib import Path

# all args are passed directly to the LineProfilerExtended constructor
@pytest.mark.line_profile.with_args(Path(__file__).parent.parent, eps=0.01)
def test_foo():
    pass
```

## Development

* Clone this repository
* Requirements:
  * [Poetry](https://python-poetry.org/)
  * Python 3.7+
* Create a virtual environment and install the dependencies

```sh
poetry install
```

* Activate the virtual environment

```sh
poetry shell
```

### Testing

```sh
pytest
```

### Releasing

Trigger the [Draft release workflow](https://github.com/utapyngo/line-profiler-extended/actions/workflows/draft_release.yml)
(press _Run workflow_). This will update the changelog & version and create a GitHub release which is in _Draft_ state.

Find the draft release from the
[GitHub releases](https://github.com/utapyngo/line-profiler-extended/releases)
and publish it. When a release is published, it'll trigger
[release](https://github.com/utapyngo/line-profiler-extended/blob/master/.github/workflows/release.yml)
workflow which creates PyPI release.

### Pre-commit

Pre-commit hooks run all the auto-formatters (e.g. `black`, `isort`), linters (e.g. `mypy`, `flake8`), and other quality
 checks to make sure the changeset is in good shape before a commit/push happens.

You can install the hooks with (runs for each commit):

```sh
pre-commit install
```

Or if you want them to run only for each push:

```sh
pre-commit install -t pre-push
```

Or if you want e.g. want to run all checks manually for all files:

```sh
pre-commit run --all-files
```

---

This project was generated using the [wolt-python-package-cookiecutter](https://github.com/woltapp/wolt-python-package-cookiecutter) template.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.0.3] - 2022-12-26
### Added
- Python 3.11 support

## [0.0.2] - 2022-11-01
### Added
- LineProfilerExtended with an IPython extension and a pytest plugin

[Unreleased]: https://github.com/utapyngo/line-profiler-extended/compare/0.0.3...master
[0.0.3]: https://github.com/utapyngo/line-profiler-extended/compare/0.0.2...0.0.3
[0.0.2]: https://github.com/utapyngo/line-profiler-extended/tree/0.0.2


tests/test_lpext.py
from pathlib import Path

import pytest
from _pytest.capture import CaptureFixture
from pytest_mock import MockerFixture

from line_profiler_extended.lpext import LineProfilerExtended
from tests import module


class TestLineProfilerExtended:
    def test_init_eps(self) -> None:
        lp = LineProfilerExtended()
        assert lp.eps == 0.1

        lp = LineProfilerExtended(eps=0.0)
        assert lp.eps == 0.0

    def test_init_exclude_regex(self, mocker: MockerFixture) -> None:
        add_modules_from_path = mocker.patch.object(
            LineProfilerExtended, "add_modules_from_path"
        )
        exclude_regex = r"^\.|conftest$|__init__|^src|test_pytest_plugin$"
        LineProfilerExtended(Path(__file__).parent.parent, exclude_regex=exclude_regex)
        add_modules_from_path.assert_called_once_with(
            Path(__file__).parent.parent, include_regex="", exclude_regex=exclude_regex
        )

    def test_init_str_path_exclude_regex(self, mocker: MockerFixture) -> None:
        add_modules_from_path = mocker.patch.object(
            LineProfilerExtended, "add_modules_from_path"
        )
        exclude_regex = r"^\.|conftest$|__init__|^src|test_pytest_plugin$"
        LineProfilerExtended("..", exclude_regex=exclude_regex)
        add_modules_from_path.assert_called_once_with(
            Path(".."), include_regex="", exclude_regex=exclude_regex
        )

    def test_init_multiple_args(self, mocker: MockerFixture) -> None:
        add_modules_from_path = mocker.patch.object(
            LineProfilerExtended, "add_modules_from_path"
        )
        add_module = mocker.patch.object(LineProfilerExtended, "add_module")
        add_function = mocker.patch.object(LineProfilerExtended, "add_function")

        LineProfilerExtended(Path(__file__).parent.parent, "..", module, module.func)
        add_module.assert_called_with(module)
        add_function.assert_called_with(module.func)
        add_modules_from_path.assert_any_call(
            Path(".."), include_regex="", exclude_regex=r""
        )
        add_modules_from_path.assert_any_call(
            Path(__file__).parent.parent, include_regex="", exclude_regex=r""
        )

    def test_init_unsupported_type(self) -> None:
        with pytest.raises(TypeError):
            LineProfilerExtended(1)  # type: ignore

    def test_add_modules_from_path(self) -> None:
        lp = LineProfilerExtended()
        assert (
            lp.add_modules_from_path(
                Path(__file__).parent.parent, exclude_regex=r"^\.|ipython"
            )
            > 0
        )

    def test_print_stats(self, capsys: CaptureFixture) -> None:
        lp = LineProfilerExtended(module, eps=1)
        lp.print_stats(output_unit=1)
        output = capsys.readouterr().out
        assert "Timer unit: 1 s\n" in output


tests/test_utils.py
import sys
from functools import wraps
from pathlib import Path
from typing import Any
from typing import Callable

from line_profiler_extended.lpext import LineProfilerExtended
from line_profiler_extended.utils import get_functions_from_module
from line_profiler_extended.utils import get_modules_from_path
from line_profiler_extended.utils import get_stats

from .module import func


def decorator(f: Callable) -> Callable:
    @wraps(f)
    def wrapper(*args: Any, **kwargs: Any) -> Any:
        return f(*args, **kwargs)

    return wrapper


class SomeClass:
    def method(self) -> None:
        pass

    @classmethod
    @decorator
    def class_method(cls) -> None:
        pass

    @staticmethod
    @decorator
    def static_method() -> None:
        pass

    @decorator
    def decorated_method(self) -> None:
        pass


def test_get_functions_from_module() -> None:
    func_names = [
        f.__qualname__ for f in get_functions_from_module(sys.modules[__name__])
    ]
    assert func.__qualname__ in func_names
    assert SomeClass.method.__qualname__ in func_names
    assert SomeClass.class_method.__qualname__ in func_names
    assert SomeClass.static_method.__qualname__ in func_names
    assert SomeClass.decorated_method.__qualname__ in func_names


def test_get_modules_from_path() -> None:
    modules = get_modules_from_path(
        Path(__file__).parent.parent, include_regex=r"test_utils$"
    )
    assert list(modules) == [sys.modules[__name__]]


def test_get_stats() -> None:
    lp = LineProfilerExtended(eps=0.0)
    lp.add_function(func)
    stats = get_stats(lp)
    print(stats)
    assert stats.startswith("Timer unit: 1e-09 s\n\nTotal time: 0 s\nFile: ")
    assert (
        "Function: func at line 1\n"
        "\n"
        "Line #      Hits         Time  Per Hit   % Time  Line Contents\n"
        "==============================================================\n"
        "     1                                           def func() -> None:\n"
        "     2                                               pass\n"
        "\n"
    ) in stats


tests/test_pytest_plugin.py
import pytest
from _pytest.pytester import Pytester


@pytest.mark.order("last")
def test_mark(pytester: Pytester) -> None:
    pytester.makepyfile(
        """
        import pytest

        import module1
        from module2 import func2

        @pytest.mark.line_profile.with_args(module1, func2, eps=0.0001)
        def test_mark() -> None:
            module1.func()
            func2()
        """,
        module1="""
        import time
        def func() -> None:
            time.sleep(0.01)
        """,
        module2="""
        import time
        def func2() -> None:
            time.sleep(0.01)
        """,
    )
    result = pytester.runpytest()
    print(result.stdout.str())
    result.stdout.fnmatch_lines(
        [
            "* Line profile result for test_mark.py::test_mark *",
            "Timer unit: 1e-09 s",
        ]
    )
    result.stdout.fnmatch_lines(
        [
            "Function: func at line 2",
            "Line #      Hits         Time  Per Hit   % Time  Line Contents",
            "==============================================================",
            "     2                                           def func() -> None:",
        ]
    )
    result.stdout.fnmatch_lines(
        [
            "Function: func2 at line 2",
            "Line #      Hits         Time  Per Hit   % Time  Line Contents",
            "==============================================================",
            "     2                                           def func2() -> None:",
        ]
    )
    result.assert_outcomes(passed=1)


src/line_profiler_extended/lpext.py
import logging
import sys
from pathlib import Path
from types import FunctionType
from types import ModuleType
from typing import Any
from typing import Callable
from typing import Dict
from typing import Optional
from typing import TextIO
from typing import Union

import line_profiler

from line_profiler_extended.utils import get_functions_from_module
from line_profiler_extended.utils import get_modules_from_path

logger = logging.getLogger(__name__)


class LineProfilerExtended(line_profiler.LineProfiler):
    def __init__(
        self,
        *items: Union[Path, ModuleType, Callable, str],
        eps: Optional[float] = None,
        include_regex: str = "",
        exclude_regex: Optional[str] = None,
    ):
        super().__init__()
        self.eps = eps if eps is not None else 0.1
        if exclude_regex is None:
            exclude_regex = r""
        for item in items:
            if isinstance(item, Path):
                self.add_modules_from_path(
                    item, include_regex=include_regex, exclude_regex=exclude_regex
                )
            elif isinstance(item, str):
                self.add_modules_from_path(
                    Path(item), include_regex=include_regex, exclude_regex=exclude_regex
                )
            elif isinstance(item, ModuleType):
                self.add_module(item)
            elif isinstance(item, FunctionType):
                self.add_function(item)
            else:
                raise TypeError(f"Unsupported argument type: {type(item)}")

    def add_function(self, func: Callable) -> int:
        if hasattr(func, "__wrapped__"):
            super().add_function(func.__wrapped__)
        else:
            super().add_function(func)
        return 1

    def add_module(self, mod: ModuleType) -> int:
        nfuncsadded = 0
        for func in get_functions_from_module(mod):
            try:
                nfuncsadded += self.add_function(func)
            except ValueError as e:  # pragma: no cover
                if e.args[0] == "code: co_code is malformed":
                    logger.warning("Skipping function %s", func.__qualname__)
                else:
                    raise
        return nfuncsadded

    def add_modules_from_path(
        self, path: Path, include_regex: str = "", exclude_regex: str = ""
    ) -> int:
        nfuncsadded = 0
        for module in get_modules_from_path(path, include_regex, exclude_regex):
            nfuncsadded += self.add_module(module)
        return nfuncsadded

    def print_stats(
        self,
        stream: Optional[TextIO] = None,
        output_unit: Optional[Union[int, float]] = None,
        **kwargs: Dict[str, Any],
    ) -> None:
        stats = self.get_stats()
        if stream is None:
            stream = sys.stdout
        if output_unit is not None:
            stream.write("Timer unit: %g s\n\n" % output_unit)
        else:
            stream.write("Timer unit: %g s\n\n" % stats.unit)

        for (fn, lineno, name), timings in sorted(
            stats.timings.items(), key=lambda x: float(sum(t[2] for t in x[1]))
        ):
            if sum(t[2] for t in timings) * stats.unit < self.eps:
                continue
            line_profiler.show_func(
                fn,
                lineno,
                name,
                timings,
                stats.unit,
                output_unit=output_unit,
                stream=stream,
                **kwargs,
            )


src/line_profiler_extended/utils.py
import importlib
import io
import os
import re
from inspect import isclass
from inspect import isfunction
from pathlib import Path
from types import ModuleType
from typing import Callable
from typing import Iterator

from line_profiler import LineProfiler


def get_modules_from_path(
    path: Path, include_regex: str = "", exclude_regex: str = ""
) -> Iterator[ModuleType]:
    cwd = os.getcwd()
    os.chdir(path)
    for p in path.glob("**/*.py"):
        module_name = str(p.relative_to(path).with_suffix("")).replace("/", ".")
        if module_name.startswith("."):
            continue
        if include_regex and not re.search(include_regex, module_name):
            continue
        if exclude_regex and re.search(exclude_regex, module_name):
            continue
        yield importlib.import_module(module_name)
    os.chdir(cwd)


def get_functions_from_module(mod: ModuleType) -> Iterator[Callable]:
    for item in mod.__dict__.values():
        if isclass(item):
            for v in item.__dict__.values():
                if isfunction(v):
                    yield v
                elif hasattr(v, "__func__") and isfunction(v.__func__):
                    yield v.__func__
        elif isfunction(item):
            yield item


def get_stats(lp: LineProfiler) -> str:
    stream = io.StringIO()
    lp.print_stats(stream=stream)
    return stream.getvalue()


src/line_profiler_extended/pytest_plugin.py
from typing import Dict
from typing import Optional

from _pytest.config import Config
from _pytest.mark import Mark
from _pytest.python import Function
from _pytest.terminal import TerminalReporter

from line_profiler_extended.lpext import LineProfilerExtended
from line_profiler_extended.utils import get_stats


def pytest_runtest_call(item: Function) -> None:
    marker: Optional[Mark] = item.get_closest_marker("line_profile")
    if not marker:
        return
    lp = LineProfilerExtended(*marker.args, **marker.kwargs)
    item_runtest = item.runtest

    # noinspection PyProtectedMember
    def runtest() -> None:
        lp.runcall(item_runtest)
        stats: Dict[str, str] = getattr(item.config, "_line_profile", {})
        stats[item.nodeid] = get_stats(lp)
        item.config._line_profile = stats  # type: ignore

    item.runtest = runtest  # type: ignore


def pytest_terminal_summary(terminalreporter: TerminalReporter, config: Config) -> None:
    reports: Dict[str, str] = getattr(config, "_line_profile", {})
    for k, v in reports.items():
        terminalreporter.write_sep("=", f"Line profile result for {k}")
        terminalreporter.write(v)


def pytest_configure(config: Config) -> None:
    config.addinivalue_line("markers", "line_profile: Line profile this test")


src/line_profiler_extended/ipython_extension.py
from io import StringIO
from typing import Any
from typing import List
from typing import Optional
from typing import Tuple

from IPython import InteractiveShell
from IPython.core.error import UsageError
from IPython.core.magic import Magics
from IPython.core.magic import line_magic
from IPython.core.magic import magics_class
from IPython.core.page import page
from IPython.utils.ipstruct import Struct

from .lpext import LineProfilerExtended


@magics_class
class LineProfilerMagics(Magics):
    @line_magic
    def lpext(self, line: str = "") -> Optional[LineProfilerExtended]:
        """Execute a statement under the line-by-line profiler from the
        line_profiler module.

        Usage:
        %lpext -p "path" -p pathlib.Path("path") -p module -p func --eps 0.01 --include .* --exclude venv <statement>

        The given statement (which doesn't require quote marks) is run via the
        LineProfiler. Profiling is enabled for the functions specified by the -f
        options. The statistics will be shown side-by-side with the code through the
        pager once the statement has completed.

        Options:

        -p <path>: profile all functions found in the given path.
        -p <module>: profile all functions found in the given module.
        -p <function>: profile the given function.
        Multiple -p options may be used. The argument may be any expression that gives
        a Python function, method, module or path (either pathlib.Path or str).
        However, one must be careful to avoid
        spaces that may confuse the option parser.

        One or more -p option is required to get any useful results.

        -eps <float>: specify the minimum execution time in seconds to be considered
        for profiling. Default is 0.1.

        -include <regex>: specify a regex to include modules in profiling.
        All modules are included by default.

        -exclude <regex>: specify a regex to exclude modules from profiling.
        Default value: "^\\.", i.e. exclude all modules starting with a dot.

        -D <filename>: dump the raw statistics out to a pickle file on disk. The
        usual extension for this is ".lprof". These statistics may be viewed later
        by running line_profiler.py as a script.

        -T <filename>: dump the text-formatted statistics with the code side-by-side
        out to a text file.

        -r: return the LineProfiler object after it has completed profiling.

        -u: specify time unit for the print-out in seconds.
        """

        # Escape quote markers.
        opts_def = Struct(
            D=[""], T=[""], p=[], u=None, eps=[None], include=[""], exclude=[None]
        )
        line = line.replace('"', r"\"").replace("'", r"\'")
        opts, arg_str = self.parse_options(
            line, "rp:D:T:u:", "eps=", "include=", "exclude=", list_all=True
        )
        opts.merge(opts_def)

        eps = None
        if opts.eps[0] is not None:
            eps = float(opts.eps[0])

        profile = LineProfilerExtended(
            *self._eval_args(opts.p),
            eps=eps,
            include_regex=opts.include[0],
            exclude_regex=opts.exclude[0],
        )

        if opts.u is not None:
            try:
                output_unit = float(opts.u[0])
            except Exception:
                raise TypeError("Timer unit setting must be a float.")
        else:
            output_unit = None

        message, output = self._run(arg_str, output_unit, profile)

        dump_file = opts.D[0]
        if dump_file:
            profile.dump_stats(dump_file)
            print(f"\n*** Profile stats pickled to file {dump_file!r}. {message}")

        text_file = opts.T[0]
        if text_file:
            pfile = open(text_file, "w")
            pfile.write(output)
            pfile.close()
            print(f"\n*** Profile printout saved to text file {text_file!r}. {message}")

        return_value = None
        if "r" in opts:
            return_value = profile

        return return_value

    def _eval_args(self, p: List[str]) -> List[Any]:
        global_ns = self.shell.user_global_ns
        local_ns = self.shell.user_ns
        args = []
        for name in p:
            try:
                args.append(eval(name, global_ns, local_ns))
            except Exception as e:
                raise UsageError(
                    f"Could not eval path, module, or function {name}.\n{e.__class__.__name__}: {e}"
                )
        return args

    def _run(
        self, arg_str: str, output_unit: Optional[float], profile: LineProfilerExtended
    ) -> Tuple[str, str]:
        global_ns = self.shell.user_global_ns
        local_ns = self.shell.user_ns

        # Add the profiler to the builtins for @profile.
        import builtins

        if "profile" in builtins.__dict__:
            had_profile = True
            old_profile = builtins.__dict__["profile"]
        else:
            had_profile = False
            old_profile = None
        builtins.__dict__["profile"] = profile
        try:
            try:
                profile.runctx(arg_str, global_ns, local_ns)
                message = ""
            except SystemExit:
                message = """*** SystemExit exception caught in code being profiled."""
            except KeyboardInterrupt:
                message = (
                    "*** KeyboardInterrupt exception caught in code being " "profiled."
                )
        finally:
            if had_profile:
                builtins.__dict__["profile"] = old_profile
        # Trap text output.
        stdout_trap = StringIO()
        profile.print_stats(stdout_trap, output_unit=output_unit)
        output = stdout_trap.getvalue()
        output = output.rstrip()
        page(output)
        print(message, end="")
        return message, output


def load_ipython_extension(ipython: InteractiveShell) -> None:
    ipython.register_magics(LineProfilerMagics)
